core/string/src/lib.rs

```diff
@@ -1144,3 +1145,348 @@ impl ToStringEscaped for [u16] {
+    const MIN_NON_ZERO_CAP: usize = 8 / Self::DATA_SIZE;
+    const IS_LATIN: bool = Self::DATA_SIZE == std::mem::size_of::<u8>();
+
+    /// Constant default for [`JsStringBuilder`]
+    pub const NEW: Self = Self {
+        cap: 0,
+        len: 0,
+        inner: NonNull::dangling(),
+        phantom_data: PhantomData,
+    };
+
+    /// Create a new `JsStringBuilder` with capacity of zero.
+    #[inline]
+    #[must_use]
+    pub const fn new() -> Self {
+        Self::NEW
+    }
+
+    /// Returns the number of elements in the inner `RawJsString`, also referred to
+    #[inline]
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Forces the length of the [`JsStringBuilder`] to `new_len`.
+    ///
+    /// # Safety
+    ///
+    /// - `new_len` must be less than or equal to `capacity()`.
+    /// - The elements at `old_len..new_len` must be initialized.
+    ///
+    #[inline]
+    pub unsafe fn set_len(&mut self, new_len: usize) {
+        debug_assert!(new_len <= self.capacity());
+
+        self.len = new_len;
+    }
+
+    /// Returns the total number of elements the inner `RawJsString` can hold without reallocating
+    #[inline]
+    #[must_use]
+    pub fn capacity(&self) -> usize {
+        self.cap
+    }
+
+    /// Returns the initalized bits of inner `RawJsString`.
+    #[must_use]
+    fn initialized_bits(&self) -> usize {
+        DATA_OFFSET + self.len() * Self::DATA_SIZE
+    }
+
+    /// create a new `JsStringBuilder` with specific capacity
+    #[must_use]
+    pub fn with_capacity(cap: usize) -> Self {
+        let layout = Self::new_layout(cap);
+        #[allow(clippy::cast_ptr_alignment)]
+        // SAFETY:
+        // The layout size of `RawJsString` is never zero, since it has to store
+        // the length of the string and the reference count.
+        let ptr = unsafe { alloc(layout) };
+
+        let Some(ptr) = NonNull::new(ptr.cast()) else {
+            std::alloc::handle_alloc_error(layout)
+        };
+        Self {
+            cap,
+            len: 0,
+            inner: ptr,
+            phantom_data: PhantomData,
+        }
+    }
+
+    /// Checks if inner `RawJsString` is allocated.
+    fn is_dangling(&self) -> bool {
+        self.inner == NonNull::dangling()
+    }
+
+    /// Returns the current inner `RawJsString` layout.
+    #[must_use]
+    fn current_layout(&self) -> Layout {
+        // SAFETY:
+        // All the checks for the validity of the layout have already been made on `new_layout`,
+        // so we can skip the unwrap.
+        unsafe {
+            Layout::for_value(self.inner.as_ref())
+                .extend(Layout::array::<T>(self.capacity()).unwrap_unchecked())
+                .unwrap_unchecked()
+                .0
+                .pad_to_align()
+        }
+    }
+
+    /// Returns the pointer of `data` of inner `RawJsString`.
+    /// # Safety
+    /// Caller should ensure that the inner is allocated.
```

Comment:
```suggestion
    /// Returns the pointer of `data` of inner `RawJsString`.
    ///
    /// # Safety
    ///
    /// Caller should ensure that the inner is allocated.
```

Minor formatting issues.

Reply:
Done.